lib/helpbase.gi

```diff
@@ -186,36 +190,53 @@ for pattern in TRANSATL do
   fi;
 od;
 
-if Length(positions) = 0 then # no matches - return immeditely
-  return [ topic ];
-fi;  
+if Length(positions) > 0 then # matches found
+  # sort data about matches accordingly to their positions in `topic'.
+  SortParallel( positions, patterns );
+
+  # Now chop the string 'topic' into a list of lists, each of them either
+  # a list of all variants from the respective spelling pattern or just
+  # a one-element list with the "glueing" string between two patterns or
+  # a pattern and the beginning or end of the string.
+
+  chop:=[];
+  begin:=1;
+  for i in [1..Length(positions)] do
+    Add( chop, [ topic{[begin..patterns[i].start-1 ]} ] );
+    Add( chop, patterns[i].pattern );
+    begin := Minimum( patterns[i].finish+1, Length(topic) );
+  od;
 
-# sort data about matches accordingly to their positions in `topic'.
-SortParallel( positions, patterns );
+  if begin < Length( topic ) then
+    Add( chop, [ topic{[begin..Length(topic)]} ] );
+  fi;
 
-# Now chop the string 'topic' into a list of lists, each of them either
-# a list of all variants from the respective spelling pattern or just
-# a one-element list with the "glueing" string between two pattersn or
-# a pattern and the beginning or end of the string. 
+  # Take the cartesian product of 'chop' and form spelling suggestions
+  # as concatenations of its elements.
 
-chop:=[];
-begin:=1;
-for i in [1..Length(positions)] do
-  Add( chop, [ topic{[begin..patterns[i].start-1 ]} ] );
-  Add( chop, patterns[i].pattern );
-  begin := Minimum( patterns[i].finish+1, Length(topic) );
-od;
+  topics := List( Cartesian(chop), Concatenation );
+
+else # no matches
+
+  topics := [ topic ];
 
-if begin < Length( topic ) then
-  Add( chop, [ topic{[begin..Length(topic)]} ] );
 fi;
 
-# Take the cartesian product of 'chop' and form spelling suggestions 
-# as concatenations of its elements. 
+r := [];
+
+for topic in topics do
+  if Length(topic) > 4 and topic{[1..3]} in [ "has" , "set" ] and topic[4]<>' ' then
+    shorttopic := topic{[4..Length(topic)]};
```

Comment:
I think there should be some comments here. One that explains what's going on ("ensure HasIsMapping is found in help even if only IsMapping ismdocumentd"), another which explains that `x > 4` is not an off-by-one accident,  but rather intentionally (and also what the intention is). Otherwise I am afraid somebody might "fix" it someday .